cardano-api/src/Cardano/Api/TxBody.hs

```diff
@@ -195,65 +226,147 @@ data TxBody era where
-    textEnvelopeType _ = "TxUnsignedByron"
-
-instance HasTextEnvelope (TxBody ShelleyEra) where
-    textEnvelopeType _ = "TxUnsignedShelley"
+    serialiseToCBOR (ShelleyTxBody era txbody txmetadata) =
+      case era of
+        -- Use the same serialisation impl, but at different types:
+        ShelleyBasedEraShelley -> serialiseShelleyBasedTxBody txbody txmetadata
+        ShelleyBasedEraAllegra -> serialiseShelleyBasedTxBody txbody txmetadata
+        ShelleyBasedEraMary    -> serialiseShelleyBasedTxBody txbody txmetadata
+
+    deserialiseFromCBOR _ bs =
+      case cardanoEra :: CardanoEra era of
+        ByronEra ->
+          ByronTxBody <$>
+            CBOR.decodeFullAnnotatedBytes
+              "Byron TxBody"
+              CBOR.fromCBORAnnotated
+              (LBS.fromStrict bs)
+
+        -- Use the same derialisation impl, but at different types:
+        ShelleyEra -> deserialiseShelleyBasedTxBody
+                        (ShelleyTxBody ShelleyBasedEraShelley) bs
+        AllegraEra -> deserialiseShelleyBasedTxBody
+                        (ShelleyTxBody ShelleyBasedEraAllegra) bs
+        MaryEra    -> deserialiseShelleyBasedTxBody
+                        (ShelleyTxBody ShelleyBasedEraMary) bs
+
+-- | The serialisation format for the different Shelley-based eras are not the
+-- same, but they can be handled generally with one overloaded implementation.
+--
+serialiseShelleyBasedTxBody :: forall txbody metadata.
+                                (ToCBOR txbody, ToCBOR metadata)
+                            => txbody -> Maybe metadata -> ByteString
+serialiseShelleyBasedTxBody txbody txmetadata =
+    CBOR.serializeEncoding' $
+        CBOR.encodeListLen 2
+     <> CBOR.toCBOR txbody
+     <> CBOR.encodeNullMaybe CBOR.toCBOR txmetadata
+
+deserialiseShelleyBasedTxBody :: forall txbody metadata pair.
+                                (FromCBOR (CBOR.Annotator txbody),
+                                 FromCBOR (CBOR.Annotator metadata))
+                              => (txbody -> Maybe metadata -> pair)
+                              -> ByteString
+                              -> Either CBOR.DecoderError pair
+deserialiseShelleyBasedTxBody mkTxBody bs =
+    CBOR.decodeAnnotator
+      "Shelley TxBody"
+      decodeAnnotatedPair
+      (LBS.fromStrict bs)
+  where
+    decodeAnnotatedPair :: CBOR.Decoder s (CBOR.Annotator pair)
+    decodeAnnotatedPair =  do
+      CBOR.decodeListLenOf 2
+      txbody     <- fromCBOR
+      txmetadata <- CBOR.decodeNullMaybe fromCBOR
+      return $ CBOR.Annotator $ \fbs ->
+        mkTxBody
+          (CBOR.runAnnotator txbody fbs)
+          (CBOR.runAnnotator <$> txmetadata <*> pure fbs)
+
+instance IsCardanoEra era => HasTextEnvelope (TxBody era) where
+    textEnvelopeType _ =
+      case cardanoEra :: CardanoEra era of
+        ByronEra   -> "TxUnsignedByron"
+        ShelleyEra -> "TxUnsignedShelley"
+        AllegraEra -> "TxBodyAllegra"
+        MaryEra    -> "TxBodyMary"
```

Comment:
Do we want to maintain consistent naming here?

Reply:
I was thinking that we change it to be consistent with the type names (ie `TxBody`), but keep existing ones for compatibility. I'm happy to be talked out of it.